src/Temporalio/Bridge/src/client.rs

```diff
@@ -0,0 +1,453 @@
+    let headers = if options.metadata.size == 0 {
+        None
+    } else {
+        Some(Arc::new(RwLock::new(
+            options.metadata.to_string_map_on_newlines(),
+        )))
+    };
+    let core_options: CoreClientOptions = match options.try_into() {
+        Ok(v) => v,
+        Err(err) => {
+            unsafe {
+                callback(
+                    user_data,
+                    std::ptr::null_mut(),
+                    runtime
+                        .alloc_utf8(&format!("Invalid options: {}", err))
+                        .into_raw(),
+                );
+            }
+            return;
+        }
+    };
+    // Spawn async call
+    let user_data = UserDataHandle(user_data);
+    let metric_meter = runtime.core.metric_meter().cloned();
+    runtime.core.tokio_handle().spawn(async move {
+        match core_options
+            .connect_no_namespace(metric_meter.as_ref(), headers)
+            .await
+        {
+            Ok(core) => {
+                let owned_client = Box::into_raw(Box::new(Client {
+                    runtime: runtime.clone(),
+                    core,
+                }));
+                unsafe {
+                    callback(user_data.into(), owned_client, std::ptr::null());
+                }
+            }
+            Err(err) => unsafe {
+                callback(
+                    user_data.into(),
+                    std::ptr::null_mut(),
+                    runtime
+                        .alloc_utf8(&format!("Connection failed: {}", err))
+                        .into_raw(),
+                );
+            },
+        }
+    });
+}
+
+#[no_mangle]
+pub extern "C" fn client_free(client: *mut Client) {
+    unsafe {
+        let _ = Box::from_raw(client);
+    }
+}
+
+#[repr(C)]
+pub struct RpcCallOptions {
+    service: RpcService,
+    rpc: ByteArrayRef,
+    req: ByteArrayRef,
+    retry: bool,
+    /// Metadata is <key1>\n<value1>\n<key2>\n<value2>. Metadata keys or
+    /// values cannot contain a newline within.
+    metadata: ByteArrayRef,
+    /// 0 means no timeout
+    timeout_millis: u32,
+    cancellation_token: *const CancellationToken,
+}
+
+// Expected to outlive all async calls that use it
+unsafe impl Send for RpcCallOptions {}
+unsafe impl Sync for RpcCallOptions {}
+
+#[repr(C)]
+pub enum RpcService {
+    Workflow = 1,
+    Operator,
+    Test,
+    Health,
+}
+
+/// If success or fail are not null, they must be manually freed when done.
+type ClientRpcCallCallback = unsafe extern "C" fn(
+    user_data: *mut libc::c_void,
+    success: *const ByteArray,
+    fail: *const ByteArray,
+);
+
+macro_rules! service_call {
+    ($service_fn:ident, $client:ident, $options:ident, $cancel_token:ident) => {{
+        let call_future = $service_fn(&$client.core, &$options);
+        if let Some(cancel_token) = $cancel_token {
+            tokio::select! {
+                _ = cancel_token.cancelled() => Err(anyhow::anyhow!("Cancelled")),
```

Comment:
Do you want to use anyhow or better typed errors?

Reply:
Yes for sure (and on RPC errors too). I need to use smarter bridge errors here. I'll end up throwing https://learn.microsoft.com/en-us/dotnet/api/system.operationcanceledexception from lang side. I'll also write a test for this (you'll note many things like proper exceptions are not included in this initial version).